src/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs

```diff
@@ -93,45 +115,227 @@ private bool AllowHttpResponseHeader
+        }
+
+        public override string[] GetValues(string header)
+        {
+            return InnerCollection.GetValues(header);
+        }
+
+        public override string GetKey(int index)
+        {
+            return InnerCollection.GetKey(index);
+        }
+
+        public override void Clear()
+        {
+            InvalidateCachedArrays();
+            if (_innerCollection != null)
+            {
+                _innerCollection.Clear();
+            }
+        }
+
+        public override string Get(int index)
+        {
+            if (_innerCollection == null)
+            {
+                return null;
+            }
+            return _innerCollection.Get(index);
+        }
+
+        public override string Get(string name)
+        {
+            if (_innerCollection == null)
+            {
+                return null;
+            }
+            return _innerCollection.Get(name);
+        }
+
+        public void Add(HttpRequestHeader header, string value)
+        {
+            if (!AllowHttpRequestHeader)
+            {
+                throw new InvalidOperationException(SR.net_headers_req);
+            }
+            this.Add(header.GetName(), value);
+        }
+
+        public void Add(HttpResponseHeader header, string value)
+        {
+            if (!AllowHttpResponseHeader)
+            {
+                throw new InvalidOperationException(SR.net_headers_rsp);
+            }
+            if (_type == WebHeaderCollectionType.WebResponse)
+            {
+                if (value != null && value.Length > ushort.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException("value", value, string.Format(CultureInfo.InvariantCulture, SR.net_headers_toolong, ushort.MaxValue));
                 }
+            }
+            this.Add(header.GetName(), value);
+        }
+
+        public void Add(string header)
+        {
+            if (string.IsNullOrWhiteSpace(header))
+            {
+                throw new ArgumentNullException("header");
+            }
+            int colpos = header.IndexOf(':');
+            // check for badly formed header passed in
+            if (colpos < 0)
+            {
+                throw new ArgumentException(SR.net_WebHeaderMissingColon, "header");
+            }
+            string name = header.Substring(0, colpos);
+            string value = header.Substring(colpos + 1);
+            name = HttpValidationHelpers.CheckBadHeaderNameChars(name);
+            value = HttpValidationHelpers.CheckBadHeaderValueChars(value);
 
-                _entriesDictionary[name] = value;
+            if (_type == WebHeaderCollectionType.WebResponse)
+            {
+                if (value != null && value.Length > ushort.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException("value", value, string.Format(CultureInfo.InvariantCulture, SR.net_headers_toolong, ushort.MaxValue));
+                }
+            }
```

Comment:
@stephentoub NormalizeCommonHeaders is an optimization that we don't  have in netcore implementation. Desktop treats common headers ( about 6 of them) as special and do quite a bit of magic to make accessing them faster( i guess that was the intention) . NormalizeCommonHeaders  move these common header to the collection when first used. Bringing CommonHeader  logic means the common header has to be treated specially in all methods. I don't know what was the reasoning for common headers to be different on desktop.